kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,614 @@
+        BaseType::SetUpSystem(rModelPart);
+        if (mHasConstraints)
+        {
+            mpConstraintVector = rCurrentProcessInfo.GetValue(CONSTRAINTS_CONTAINER);
+            Constraints_SetUp(rModelPart, mpConstraintVector);
+        }
+    }
+
+    /** Destructor.
+	 */
+    virtual ~ResidualBasedBlockBuilderAndSolverWithConstraints()
+    {
+    }
+
+    void BuildAndSolve(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart &rModelPart,
+        TSystemMatrixType &A,
+        TSystemVectorType &Dx,
+        TSystemVectorType &b) override
+    {
+        KRATOS_TRY
+
+        Timer::Start("Build");
+
+        if (mHasConstraints)
+        {
+            Constraints_ExecuteBeforeBuilding(rModelPart, mpConstraintVector);
+            Build(pScheme, rModelPart, A, b);
+        }
+        else
+        {
+            BaseType::Build(pScheme, rModelPart, A, b);
+        }
+
+        Timer::Stop("Build");
+
+        this->ApplyDirichletConditions(pScheme, rModelPart, A, Dx, b);
+
+        if (mHasConstraints)
+            Constraints_ExecuteAfterBuilding(rModelPart, mpConstraintVector);
+
+        if (this->GetEchoLevel() == 3)
+        {
+            std::cout << "before the solution of the system" << std::endl;
+            std::cout << "System Matrix = " << A << std::endl;
+            std::cout << "unknowns vector = " << Dx << std::endl;
+            std::cout << "RHS vector = " << b << std::endl;
+        }
+
+        double start_solve = OpenMPUtils::GetCurrentTime();
+        Timer::Start("Solve");
+        if (mHasConstraints)
+            Constraints_ExecuteBeforeSolving(A, Dx, b, rModelPart, mpConstraintVector);
+
+        this->SystemSolveWithPhysics(A, Dx, b, rModelPart);
+
+        Timer::Stop("Solve");
+        double stop_solve = OpenMPUtils::GetCurrentTime();
+        if (this->GetEchoLevel() >= 1 && rModelPart.GetCommunicator().MyPID() == 0)
+            std::cout << "system solve time: " << stop_solve - start_solve << std::endl;
+
+        if (this->GetEchoLevel() == 3)
+        {
+            std::cout << "after the solution of the system" << std::endl;
+            std::cout << "System Matrix = " << A << std::endl;
+            std::cout << "unknowns vector = " << Dx << std::endl;
+            std::cout << "RHS vector = " << b << std::endl;
+        }
+        if (mHasConstraints)
+            Constraints_ExecuteAfterSolving(A, Dx, b, rModelPart, mpConstraintVector);
+
+        KRATOS_CATCH("")
+    }
+
+    //**************************************************************************
+    //**************************************************************************
+    // This is modified to include the MPC information.
+    //
+    void Build(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart &rModelPart,
+        TSystemMatrixType &A,
+        TSystemVectorType &b) override
+    {
+        KRATOS_TRY
+        if (!pScheme)
```

Comment:
`KRATOS_ERROR_IF_NOT(pScheme)`

Reply:
also here .. this is also same in base builder and sovler. I will change this here but we have to make sure it is also changed in the base builder and solver. 

This should be done in separate pull request.